libsolidity/ast/Types.cpp

```diff
@@ -199,18 +148,78 @@ util::Result<TypePointers> transformParametersToExternal(TypePointers const& _pa
 	return transformed;
 }
 
+bool compareTypes(Type const* lhs, Type const* rhs)
+{
+	solAssert(lhs && rhs, "");
+	return lhs->richIdentifier() < rhs->richIdentifier();
 }
+using TypeSet = std::set<Type const*, decltype(compareTypes)*> ;
 
-vector<frontend::Type const*> solidity::frontend::oversizedSubtypes(frontend::Type const& _type)
+void storageOversizedSubtypesInner(
+	Type const* _type,
+	bool _includeType,
+	set<string>& _seen,
+	TypeSet& _oversizedSubtypes
+)
 {
-	set<StructDefinition const*> structsSeen;
-	TypeSet oversized;
-	oversizedSubtypesInner(_type, false, structsSeen, oversized);
-	vector<frontend::Type const*> res;
+	for (Type const* subtype: _type->subtypes())
+	{
+		string id = subtype->richIdentifier();
+		if (_seen.count(id))
+			continue;
+		_seen.insert(id);
+		bool includeType = _type->category() != Type::Category::Struct;
+		storageOversizedSubtypesInner(subtype, includeType, _seen, _oversizedSubtypes);
+		_seen.erase(id);
+	}
+
+	if (_includeType && _type->storageSizeUpperBound() >= bigint(1) << 64)
+		_oversizedSubtypes.insert(_type);
+}
+
+}
+
+vector<Type const*> Type::storageOversizedSubtypes() const
+{
+	set<string> seen;
+	TypeSet oversized(compareTypes);
+	storageOversizedSubtypesInner(this, false, seen, oversized);
+	vector<Type const*> res;
 	copy(oversized.cbegin(), oversized.cend(), back_inserter(res));
 	return res;
 }
 
+BoolResult Type::validForLocation(DataLocation _loc, set<string>* _seen) const
+{
+	if (_seen == nullptr)
+	{
+		set<string> seen;
+		return validForLocation(_loc, &seen);
+	}
+
+	for (Type const* subtype: subtypes())
+	{
+		string id = subtype->richIdentifier();
+		if (_seen->count(id))
+			continue;
+		_seen->insert(id);
+		BoolResult result = subtype->validForLocation(_loc, _seen);
+		_seen->erase(id);
+		if (!result)
+			return result;
+	}
+
+	bigint size = sizeUpperBound(_loc);
```

Comment:
> It would be more efficient to combine them into one function that both returns the size and at the same time says if the type is valid.

The same `sizeUpperBound` function is used for collision warnings, see `storageOversizedSubtypes`. And in that case we do recursion to find all oversized subtypes.

I do agree though that the function is inefficient and could be optimized.

Reply:
Well, ok then. I see that `sizeUpperBound()` isn't fully recursive anyway for memory and calldata. I think we should refactor it later to be fully recursive.

Then the collision check could also be reduced to checking the size of the top-level type instead of havng to check every level separately.

Reply:
> Then the collision check could also be reduced to checking the size of the top-level type instead of havng to check every level separately.

I thought about it too. Maybe this is something to clarify first. If we limit collision checks to top-level types only, the code will get significantly simpler. But we would also decrease diagnostic details.

Reply:
Right. But we could still optimize it by only descending down the tree if the top-level type is actually oversized. And that's something that could be kept in static analyzer - so that it's only done if we do need the diagnostics.

You could make it both simple and reusable by just providing "building blocks" for the checks instead of functions like `storageOversizedSubtypes()`. I.e.:
- Define proper, recursive `sizeUpperBound()`
- Define a function to recursively list all the subtypes (that I proposed in [another comment](https://github.com/ethereum/solidity/pull/9746/files#r484905984)).

Then if would be easy to reuse these parts for both collision checks and size limit checks. I.e. static analyzer would check the total size and, if it's oversized, just iterate over the subtypes and print the oversized ones. It could also sort them according to its own criteria to ensure deterministic messages.